src/ranges.rs

```diff
@@ -0,0 +1,48 @@
+use syntax::ast::*;
+use syntax::codemap::Spanned;
+use rustc::lint::{Context, LintArray, LintPass};
+
+declare_lint! {
+    pub RANGE_STEP_BY_ZERO, Warn,
+    "Warn on uses of Range::step_by(0)"
+}
+
+#[derive(Copy,Clone)]
+pub struct StepByZero;
+
+impl LintPass for StepByZero {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(RANGE_STEP_BY_ZERO)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprMethodCall(Spanned { node: ref ident, .. }, _,
+                              ref args) = expr.node {
+            // Only warn on literal ranges.
+            if ident.name.as_str() == "step_by" && args.len() == 2 &&
+                is_lit_range(&args[0]) && is_lit_zero(&args[1]) {
+                cx.span_lint(RANGE_STEP_BY_ZERO, expr.span,
+                             "Range::step_by(0) may loop forever.")
```

Comment:
Hah, I actually thought about the trailing period but apparently my fingers typed it of their own volition. :)

I said "may" here because you could use `.take(_)` or similar to avoid an infinite loop. And the lint isn't checking that one way or another.